### ISSUE TITLE
Import syntax for package and child module is ambigious

### ISSUE THREAD

**Issue by jaraco** [python/cpython#104453]
When a user is importing a package and its child and intends to use attributes of both, the syntax is ambiguous. For example:

```
import foo
import foo.bar

foo.baz()
foo.bar.bing()
```

Because `import foo.bar` binds `foo` in the namespace and ensures `foo.bar` is imported, the preceding `import foo` is superfluous (a subset of the effect of the following line). I'll call this form "duplicative" and the alternative "consolidated" (avoiding terms like implicit/explicit to avoid zen biasing):

```
import foo.bar

foo.baz()
foo.bar.bing()
```

Although both forms are allowed, it creates an unnecessary ambiguity as to which form is preferred. This ambiguity has directly affected my productivity, serving as the sole blocker on a review of a proposed change.

It would be advantageous for Python to provide a recommended style on which linters and other style guides could base their recommendations if for no other reason than to create the "preferably one way" to approach this need.

## tradeoffs

There are advantages and disadvantages to the different forms. I'll capture a few here.

Advantages of the duplicative form is that it provides the author with the ability to signal that both "foo" and "foo.bar" are expected to be used in the module. If adopted broadly, linters would be able to use that signal to detect "unused" imports of the top level or "missing duplicative imports".

Advantages of the consolidated form is that it provides a more concise syntax and aligns with common usage (including an [explicit recommendation that `os.path` be accessed from `import os`](https://github.com/python/cpython/blob/46f1c78eebe08e96ed29d364b1804dd37364831d/Lib/posixpath.py#L3-L4)).

## existing usage

In a quick review of the code, looking at `importlib` and its submodule imports, the dominant usage is to use the consolidated form, with some examples of the duplicative form.

## explicit poll

In a very informal poll in Discord, Python core devs showed a split opinion on the matter with maybe a preference for the duplicative form (7 vs. 4.5).

![image](https://github.com/python/cpython/assets/308610/b33ab3f1-813a-480e-bcc0-c9f21a99eedc)

## proposal

I propose that we update either the Python documentation (probably [here](https://docs.python.org/3/reference/simple_stmts.html#the-import-statement)) to give preferred guidance or to update PEP 8 to include such guidance. The guidance should be for the duplicative form unless there's substantial dissent.

**Comment by terryjreedy** [comment]
I was initially almost surprised that `import foo.bar; foo.baz()` worked because I have almost never seen this shortened sequence.  (Of course, upon a moment's thought, binding 'foo' to 'sys.modules["foo"]' is the only sensible result of the import.)  Look at the first 65 lines of IDLE's [editor.py]( https://github.com/python/cpython/blob/main/Lib/idlelib/editor.py) to see what is routine in idlelib.  (The `from` forms for tkinter are also normal in my experience.)

It makes me sad that a dispute over this would stop an otherwise good merge.  But if a choice were to be made (I don't consider doing so to be mandatory), I prefer requiring the initial 'import foo'.  "Import foo and bind it to 'foo'; then import foo.bar and bind it to 'bar' in foo." is the intention. To me, it would not have been completely crazy to require this.

**Comment by CAM-Gerlach** [comment]
> I'll call this form "duplicative" and the alternative "consolidated" (avoiding terms like implicit/explicit to avoid zen biasing):

I would view those terms are _more_ clearly biased with a stronger connotation (negative for the former, somewhat positive for the latter) than explicit/implicit, which more neutrally and conventionally contrasts the two and where the claimed bias there comes from the one's opinions/principles for the language in question, which is in fact of key relevance here.

> I propose that we update either the Python documentation (probably [here](https://docs.python.org/3/reference/simple_stmts.html#the-import-statement)) to give preferred guidance or to update PEP 8 to include such guidance.

IMO, the authoritative language specification reference is not the right place for a mere stylistic preference; PEP 8 (PEP-8) would be the well-established place for that.

**Comment by gvanrossum** [comment]
Yeah, we should require both. The reasoning being that if you later remove the call to foo.bar.baz() you should be able to just delete that import, regardless of whether there is still an foo.baz() call some where.

**Comment by warsaw** [comment]
+1 with @gvanrossum - also, the `os.path` example isn't a good one since it's long been a weird special case.
